backend/monoprocess/monoprocess.go

```diff
@@ -0,0 +1,180 @@
+	// loop until either we find a task or the context is cancelled
+	for {
+		if w, err := b.Backend.GetWorkflowTask(ctx); w != nil || err != nil {
+			return w, err
+		}
+		b.logger.DebugContext(ctx, "worker waiting for workflow task signal")
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-b.workflowSignal:
+			b.logger.DebugContext(ctx, "worker got a workflow task signal")
+		}
+	}
+}
+
+func (b *monoprocessBackend) GetActivityTask(ctx context.Context) (*task.Activity, error) {
+	// loop until either we find a task or the context is cancelled
+	for {
+		if a, err := b.Backend.GetActivityTask(ctx); a != nil || err != nil {
+			return a, err
+		}
+		b.logger.DebugContext(ctx, "worker waiting for activity task signal")
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-b.activitySignal:
+			b.logger.DebugContext(ctx, "worker got an activity task signal")
+		}
+	}
+}
+
+func (b *monoprocessBackend) CreateWorkflowInstance(ctx context.Context, instance *workflow.Instance, event *history.Event) error {
+	if err := b.Backend.CreateWorkflowInstance(ctx, instance, event); err != nil {
+		return err
+	}
+	b.notifyWorkflowWorker(ctx)
+	return nil
+}
+
+func (b *monoprocessBackend) CompleteWorkflowTask(
+	ctx context.Context,
+	task *task.Workflow,
+	instance *workflow.Instance,
+	state core.WorkflowInstanceState,
+	executedEvents, activityEvents, timerEvents []*history.Event,
+	workflowEvents []history.WorkflowEvent,
+) error {
+	if err := b.Backend.CompleteWorkflowTask(ctx, task, instance, state, executedEvents, activityEvents, timerEvents, workflowEvents); err != nil {
+		return err
+	}
+
+	hasWorkflowTasks := false
+	hasActivityTasks := false
+
+	for _, e := range executedEvents {
+		if e.Type == history.EventType_WorkflowTaskStarted {
+			hasWorkflowTasks = true
+			break
+		}
+	}
+	for _, e := range activityEvents {
+		if e.Type == history.EventType_ActivityScheduled {
+			hasActivityTasks = true
+			break
+		}
+	}
+	for _, e := range timerEvents {
+		attr, ok := e.Attributes.(*history.TimerFiredAttributes)
+		if !ok {
+			b.logger.Warn("unknown attributes type in timer event", "type", reflect.TypeOf(e.Attributes).String())
+			continue
+		}
+		b.logger.DebugContext(ctx, "scheduling timer to notify workflow worker")
+		// Note that the worker will be notified even if the timer event gets
+		// cancelled. This is ok, because the poller will simply find no task
+		// and continue.
+		time.AfterFunc(attr.At.Sub(time.Now()), func() { b.notifyWorkflowWorker(context.Background()) })
+	}
+	for _, e := range workflowEvents {
+		if e.HistoryEvent.Type == history.EventType_WorkflowExecutionStarted ||
+			e.HistoryEvent.Type == history.EventType_SubWorkflowCompleted ||
+			e.HistoryEvent.Type == history.EventType_WorkflowExecutionCanceled {
+			hasWorkflowTasks = true
+			break
+		}
+	}
+
+	// notify workers about potential new tasks
+	if hasWorkflowTasks {
+		b.notifyWorkflowWorker(ctx)
+	}
```

Comment:
Thinking more about this, I'm wondering if we even need `hasWorkflowTasks`. To keep this simple(r), it might be enough to just always call `notifyWorkflowWorker` after `CompleteWorkflowTask` is successfully executed. There might or might not be a task but checking once should be fine.

Reply:
I'm on board with this ☝🏻 While it may have the side effect of notifying when there is no work to do, it would be safer than cherry-picking the conditions that guarantee a task to exist, conditions that may change over time and risking the implementation to become inconsistent.

Reply:
Agreed and done!